### ISSUE TITLE
How to add reveal.js plugins ?

### ISSUE THREAD

**Issue by Sigmun** [webpro/reveal-md#102]
Following the [Tool](https://github.com/hakimel/reveal.js/wiki/Plugins,-Tools-and-Hardware#tools) link provided in the Readme, I need to add some plugins in my slides (menu, charts,chalkboard...). How to proceed since reveal-md uses its own reveal.js ?

**Comment by webpro** [comment]
Currently reveal-md has no support for adding extra reveal.js plugins.

For what it's worth: these plugins are already included: https://github.com/webpro/reveal-md/blob/73d14763cbb3218d616a4c55f4eaf88341b25112/lib/template/reveal.html#L45-L53

I don't know how other (third party) plugins should be added, perhaps this feature is easy to add to reveal-md.

**Comment by JohnGuan** [comment]
@webpro Does reveal-md has support for adding extra reveal.js plugins now?

**Comment by webpro** [comment]
No, not really. There are two options:

- Copy the default template and add the plugins there (and use it with `--template`)
- Fork reveal-md (and submit a pull request) to add (and disable?) plugins.

**Comment by JohnGuan** [comment]
I tried to add blow content into `reveal.json`, but it turns plugins name to string, so it can't initialize normally.
```
{
    "plugins":[
        "RevealChalkboard",
        "RevealMenu"
        ]
}
```
I think if we use js code to config reveal.js options will be better for add, enable or disable plugins.

---

blow is `reveal-md.json` content
```
{
    "scripts": [
        "plugin/chalkboard/plugin.js",
        "plugin/reveal.js-menu/menu.js"
        ]
  }
```

**Comment by JohnGuan** [comment]
I found a way to add Three Party plugins.

First, specify the location of the plugins in the `reveal-md.json` file.
The last item 'plugin.js' is reserved for enabling plugins.
```
{
    "scripts": [
        "plugin/chalkboard/plugin.js",
        "plugin/reveal.js-menu/menu.js",
        "plugin.js"
        ]
  }
```

Then write in `plugin.js` which plugins need to be enabled
```
options.plugins.push(
    RevealChalkboard,
    RevealMenu
);
```

**Comment by Aircl0wn** [comment]
work great.
@JohnGuan do you have a way of passing on config to the plugins by any chance?

**Comment by jvanheugten** [comment]
@Aircl0wn : To add a config for a plugin, create a `reveal.json` file:

Example for the config of the [chalkboard](https://github.com/rajgoel/reveal.js-plugins/tree/master/chalkboard) plugin:
```json
{
    "chalkboard": {
        "boardmarkerWidth": 10,
        "chalkWidth": 10,
        "readOnly": false,
        "src": "test.json",
        "toggleChalkboardButton": { "left": "30px", "bottom": "30px", "top": "auto", "right": "auto" },
        "toggleNotesButton": { "left": "30px", "bottom": "30px", "top": "auto", "right": "auto" },
        "eraser": {"radius": 50, "src": "plugins/chalkboard/img/sponge.png"}
    }
}
```
Note that the options should be in json format, so all strings should have quotations.

It would be great to add this to the reveal-md documentation.

**Comment by Aircl0wn** [comment]
thnx @jvanheugten !

**Comment by webpro** [comment]
Thanks @JohnGuan and @jvanheugten! I added this to the readme the other day: https://github.com/webpro/reveal-md#scripts-preprocessors-and-plugins.

**Comment by krillin666** [comment]
Excuse me before hand if I'm missing something obvious. However, I cannot enable the TOC-progress plugin [https://github.com/e-gor/Reveal.js-TOC-Progress](https://github.com/e-gor/Reveal.js-TOC-Progress). I've enable other plugins but with this one I fail with every attempt.
I load the plugin javascript in the appropriate `reveal-md.json` file but I haven't figured out how to initialize it in the `plugin.js` file.
I've tried: 
- RevealTocprogress
- RevealToc-progress
- RevealTOC-progress
- RevealTOC_progress

Any hint ?

Thank you !

**Comment by webpro** [comment]
https://github.com/e-gor/Reveal.js-TOC-Progress/blob/master/plugin/toc-progress/toc-progress.js#L14

So that would be `toc_progress`

**Comment by krillin666** [comment]
> https://github.com/e-gor/Reveal.js-TOC-Progress/blob/master/plugin/toc-progress/toc-progress.js#L14
> 
> So that would be `toc_progress`

Hey thanks for the quick answer. However, I think I had tried that too but to no avail.

Just to be sure, my `plugin.js` is : 

```
   1   │ options.plugins.push(
   2   │     RevealMenu,
   3   │     RevealAnimate,
   4   │     RevealAnything,
   5   │     RevealChart,
   6   │     RevealMath,
   7   │     toc_progress
   8   │ );
```
And my `reveal-md.json` file is:

```
   1   │ {
   2   │     "scripts": [
   3   │         "plugin/reveal.js-menu/menu.js",
   4   │         "plugin/anything/plugin.js",
   5   │         "plugin/reveal.js-plugins/animate/plugin.js",
   6   │         "plugin/reveal.js-plugins/animate/svg.min.js",
   7   │         "plugin/chalkboard/plugin.js",
   8   │         "plugin/chart/Chart.min.js",
   9   │         "plugin/chart/plugin.js",
  10   │         "plugin/math/math.js",
  11   │         "plugin/toc-progress/toc-progress.js",
  12   │         "plugin.js"
  13   │     ]
  14   │ }
```
But when I run `reveal-md a.md --script plugin.js -w --css twocolumns.css --theme white` the TOC does not appear:

![toc1](https://user-images.githubusercontent.com/60305303/118093435-f3c61280-b3bc-11eb-9e14-5b19fdbce628.png)

![toc2](https://user-images.githubusercontent.com/60305303/118093449-f7f23000-b3bc-11eb-92fe-1cbc0c7e8b22.png)

Thank you for your time !

**Comment by webpro** [comment]
> But when I run `reveal-md a.md --script plugin.js`

The argument is `scripts` (not `script`)

**Comment by Ouch1978** [comment]
@JohnGuan 

May I know if this way still working or not?

I've tried with the following process, but cannot make the menu show up.

1. Run yarn add reveal.js-menu.
1. Create a folder named "plugin".
2. Copy reveal.js-menu from node_modules to the plugin folder.
3. Add the following content into `reveal-md.json`:
```json
    "scripts": [
        "plugin/reveal.js-menu/menu.js",
        "plugin.js"
    ]
```
4. Add a new file named `plugin.js`, and add the following content:
```json
options.plugins.push(
  RevealMenu
);
```
5. run reveal-md with command:
```shell
reveal-md slides/Test1.md --scripts plugin.js
```

Would you please tell me if I did something wrong?

Thank you.

**Comment by Aircl0wn** [comment]
@Ouch1978 : is your `scripts` block in `reveal-md.json` inside accolades?

```
{
    "scripts": [
        "plugin/reveal.js-menu/menu.js",
        "plugin.js"
    ]
}
```

**Comment by Ouch1978** [comment]
@Aircl0wn 

Yes, I did.

It's

```json
{
    "separator": "\n---\n",
    "verticalSeparator": "\n----\n",
    "scripts": [
        "plugin/reveal.js-menu/menu.js",
        "plugin.js"
    ]
}
```

**Comment by Ouch1978** [comment]
I finally figure it out.

Just execute 
```shell
reveal-md slides/Test1.md
```

without `--scripts plugin.js` is working for me.

**Comment by Aircl0wn** [comment]
Yes, that's what I'm doing as well.
Didn't think that would matter, hence why I looked elsewhere. Glad it's working.

**Comment by Thus0** [comment]
@krillin666 : did the plugin "toc_progress" worked out of the box ? The plugin seems to be loaded but the toc does not appear in the footer. I loaded "RevealMenu" plugin so I think my reveal.json, plugin.json and reveal-md.json are correct

toc_progress seems to parse <h1>, <h2> and <h3> headers so I tried to preprocess the markdown file but with no luck so far.

Anyone use also the toc_progress ?

**Comment by lukenavo** [comment]
Sorry for bother you all, but I'm not able to install chalkboard plugin in my reveal-md slide.
I have a lot of doubts...
1 - is it possible to have chalkboard in a static output slide?
2 - I've downloaded the entire plugin collection. Where must I put it? I have to install them with some command? I have put them in /plugin where there are *.md files
3 - I've create a reveal-md.json  in the same location of .md file with: 

{
    "scripts": [
        "plugin/chalkboard/plugin.js",
        "plugin.js"
        ]
  }

then I create plugin.js in the same location of .md file with:

options.plugins.push(
    RevealChalkboard,
    RevealMenu
);

but when I start the command
reveal-md name.md --static 
or 
reveal-md name.md

no chalkboard appears. What is wrong?

Thanks for any suggesion,
Luca

**Comment by lukenavo** [comment]
Sorry, plugin.js is:
options.plugins.push(
    RevealChalkboard
);

**Comment by jackyliu16** [comment]
> Sorry for bother you all, but I'm not able to install chalkboard plugin in my reveal-md slide. I have a lot of doubts... 1 - is it possible to have chalkboard in a static output slide? 2 - I've downloaded the entire plugin collection. Where must I put it? I have to install them with some command? I have put them in /plugin where there are *.md files 3 - I've create a reveal-md.json in the same location of .md file with:
> 
> { "scripts": [ "plugin/chalkboard/plugin.js", "plugin.js" ] }
> 
> then I create plugin.js in the same location of .md file with:
> 
> options.plugins.push( RevealChalkboard, RevealMenu );
> 
> but when I start the command reveal-md name.md --static or reveal-md name.md
> 
> no chalkboard appears. What is wrong?
> 
> Thanks for any suggesion, Luca

One interesting thing is that I don't have a way to cover RevealMenu in static output (but I can cover it in direct open by `reveal-md.cmd slides.md`).

even i have seem the  plugins has been include in `index.html`
```
    <script src="./_assets/./plugin/reveal.js-menu/menu.js"></script>
    <script src="./_assets/./plugin/chalkboard/plugin.js"></script>
    <script src="./_assets/./plugin/plugin.js"></script>
```
and `_assets/plugin/plugin.js`

```
options.plugins.push(
    RevealMenu          // left menu
);
```
but it couldn't display the menu buttom(by the ways i have add `console.log("sth") into this js plugins, and it could display in f12, which i guess means that this plugins has been contain in html file successed but not working[i actually not familiar with js])

_And as soon as I tried to introduce the RevealChalkboard script into my program, the entire page would not be displayed_

**Comment by MukundMohanan** [comment]
Hi,

I'm trying to add RevealChart plugin to my presentation but got an error in browser console
```
Uncaught ReferenceError: RevealChart is not defined
    at plugin.js:2:5
```
Below is my complete set-up

Project structure
```
│── plugin
│       └─ chart
│             └─ Chart.min.js
│             └─ plugin.js
│
│── src
│          └─README.md
│── plugin.js
│── reveal-md.json
│── reveal.json
```

```plugin.js```
```js
options.plugins.push(
    RevealChart,
);
```
```reveal.json```
```json
{
    "chart": {
		"defaults": {
			"color": "lightgray",
			"scale": {
				"beginAtZero": true,
				"ticks": { "stepSize": 1 },
				"grid": { "color": "lightgray" }
			}
		},
		"line": { "borderColor": [ "rgba(20,220,220,.8)" , "rgba(220,120,120,.8)", "rgba(20,120,220,.8)" ], "borderDash": [ [5,10], [0,0] ] },
		"bar": { "backgroundColor": [ "rgba(20,220,220,.8)" , "rgba(220,120,120,.8)", "rgba(20,120,220,.8)" ]},
		"pie": { "backgroundColor": [ ["rgba(0,0,0,.8)" , "rgba(220,20,20,.8)", "rgba(20,220,20,.8)", "rgba(220,220,20,.8)", "rgba(20,20,220,.8)"] ]}
	}
}
```
```reveal-md.json```
```json
{
    "separator": "^\n\n\n",
    "verticalSeparator": "^\n\n",
    "scripts": [
        "plugin/chart/Chart.min.js",
        "plugin/chart/plugin.js",
        "plugin.js"
    ]
}
```
```plugin/chart/plugin.js``` has code from [chart plugin from reveal.js-plugins](https://github.com/rajgoel/reveal.js-plugins/blob/master/chart/plugin.js)

```plugin/chart/Chart.min.js``` has code from [char min js from cdn](https://cdnjs.cloudflare.com/ajax/libs/Chart.js/4.2.1/chart.min.js)

and I run ```reveal-md src/README.md --scripts plugin.js```.

I'm stuck with the error in browser console saying 
``` Uncaught ReferenceError: RevealChart is not defined
    at plugin.js:2:5
```

**Comment by webpro** [comment]
In general, some issues seem to come from the fact that the CLI config (`--scripts`) has higher priority than the local config in `reveal.json` and overrides the latter completely (it does not merge the items in the array or something). So you should use either the cli argument `--scripts` OR the JSON config `"scripts": []`.

What also may help is to look in the resulting HTML (use the inspector in the browser) and look for the `<script>` tags near the end of the `<body>`.

**Comment by Silvanoc** [comment]
I've been using in the past the Reveal.JS plugin [reveal-plantuml](https://github.com/reveal-plantuml/reveal-plantuml.github.io) without problems on reveal-md 3.5.0.

I've updated to the latest release and it fails with the error message `Uncaught ReferenceError: RevealPlantUML is not defined` (plugin.js is trying to `push` `RevealPlantUML`). I've been able to pin down the issue to the switch from reveal-md 3.5.0 to 4.0.0. Since that version of reveal-md is also implying a major version bump in  Reveal.JS (from 3.9.2 to 4.0.0) and the plugin last release happened between that major version switch, I suspect that it's somehow incompatible.

Now my question is, could anybody give me a hint on how I could try to fix it myself? I'm a newbie on JavaScript all all its tooling and frameworks.
